Module3HM3_Algorithms/SearchingSortingRecursion/CollectionExtension.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SearchingSortingRecursion
+{
+    public static class CollectionExtension
+    {
+        public static int BinarySearch<T>(this IEnumerable<T> collection, T item) where T : IComparable<T>
+        {
+            if (collection == null) throw new ArgumentNullException("collection");
+            if (item == null) throw new ArgumentNullException("item");
+            if (!collection.Any()) return -1;
+
+            IList<T> list = collection is IList<T> ? (IList<T>)collection : new List<T>(collection);
```

Comment:
If you want to use IList, use it instead of the IEnumerable. Here you are breaking the sense of binary search - quick search algo by simply copying N elements into a new List.